protocol/flows/src/v5/txrelay/flow.rs

```diff
@@ -0,0 +1,278 @@
+        self.start_impl().await
+    }
+}
+
+impl RelayTransactionsFlow {
+    #[allow(dead_code)]
+    pub fn new(ctx: FlowContext, router: Arc<Router>, incoming_route: IncomingRoute, msg_route: IncomingRoute) -> Self {
+        Self { ctx, router, invs_route: TwoWayIncomingRoute::new(incoming_route), msg_route }
+    }
+
+    async fn start_impl(&mut self) -> Result<(), ProtocolError> {
+        // trace!("Starting relay transactions flow with {}", self.router.identity());
+        loop {
+            // Loop over incoming block inv messages
+            let inv = self.invs_route.dequeue().await?;
+            // trace!("Receive an inv message from {} with {} transaction ids", self.router.identity(), inv.len());
+
+            // Transaction relay is disabled if the node is out of sync and thus not mining
+            if self.ctx.is_ibd_running() {
+                continue;
+            }
+
+            let requests = self.request_transactions(inv).await?;
+            self.receive_transactions(requests).await?;
+        }
+    }
+
+    async fn request_transactions(
+        &self,
+        transaction_ids: Vec<TransactionId>,
+    ) -> Result<Vec<RequestScope<TransactionId>>, ProtocolError> {
+        // Build a vector with the transaction ids unknown in the mempool and not already requested
+        // by another peer
+        let requests: Vec<_> = transaction_ids
+            .iter()
+            .filter_map(|transaction_id| {
+                if !self.is_known_transaction(transaction_id) {
+                    self.ctx.try_adding_transaction_request(transaction_id.to_owned())
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // Request the transactions
+        if !requests.is_empty() {
+            // TODO: determine if there should be a limit to the number of ids per message
+            // trace!("Send a request to {} with {} transaction ids", self.router.identity(), requests.len());
+            self.router
+                .enqueue(make_message!(
+                    Payload::RequestTransactions,
+                    RequestTransactionsMessage { ids: requests.iter().map(|x| x.req.into()).collect() }
+                ))
+                .await?;
+        }
+
+        Ok(requests)
+    }
+
+    fn is_known_transaction(&self, transaction_id: &TransactionId) -> bool {
+        // Ask the transaction memory pool if the transaction is known
+        // to it in any form (main pool or orphan).
+        self.ctx.mining_manager().get_transaction(transaction_id, true, true).is_some()
+    }
+
+    /// Returns the next Transaction or TransactionNotFound message in msg_route,
+    /// returning only one of the message types at a time.
+    ///
+    /// Populates the invs_route queue with any inv messages that meanwhile arrive.
+    async fn read_response(&mut self) -> Result<Response, ProtocolError> {
+        loop {
+            tokio::select! {
```

Comment:
There is no need to use select here, you only need to await on `self.msg_route.recv()`. The whole point of using two incoming routes (msg and invs) unlike go is that you don't need to do this. The main loop dequeues invs, and msg is used only after sending tx requests. This means that unlike block invs, you don't need the `TwoWayIncomingRoute` object here. The difference is that block invs requests can come indirectly from orphan roots, while txs don't have this mechanism

Reply:
The reason why I kept the `TwoWayIncomingRoute` design is because processing the responses to a `RequestTransactions` with full capacity (thus an `InvTransaction` holding `MAX_INV_PER_TX_INV_MSG` ids) might take a significant time. Since the `IncomingRoute` provided by `Router::subscribe` is bounded to 128, the `invs_route` queue might fill up. So using the select prevents this risk.

That being said, this risk is probably low now by with network transaction rate increasing, it might materialize in the future.

WDYT?

Reply:
The two-way mechanism simply moves the size issue to the unlimited deque, so we can simply use a larger channel if needed.

Reply:
Perhaps add a method `Router::subscribe_with_capacity` and use a larger value for invs